src/types/params.ts

```diff
@@ -96,18 +96,33 @@ export type ExtractRouteParamTypes<TRoute> = TRoute extends {
   : Record<string, unknown>
 
 /**
- * Extracts combined types of path and query parameters for a given route, creating a unified parameter object.
- * This parameter object type represents the expected type when accessing params from router.route or useRoute.
+ * Does everything that ExtractRouteParamTypes does, but takes into consideration optional properties.
+ * Differs from ExtractRouteParamTypes in that it also reads the string `value` from WithParams to determine what should be optional.
+ * @template TRoute - The route type from which to extract and merge parameter types.
+ * @returns A record of parameter names to their respective types, extracted and merged from both path and query parameters.
+ */
+export type ExtractRouteParamTypesOptionalReading<TRoute> = TRoute extends {
+  host: { params: infer HostParams extends Record<string, Param> },
+  path: { params: infer PathParams extends Record<string, Param> },
+  query: { params: infer QueryParams extends Record<string, Param> },
+  hash: { params: infer HashParams extends Record<string, Param> },
+}
+  ? Identity<MakeOptional<ExtractParamTypesOptionalReading<HostParams> & ExtractParamTypesOptionalReading<PathParams> & ExtractParamTypesOptionalReading<QueryParams> & ExtractParamTypesOptionalReading<HashParams>>>
+  : Record<string, unknown>
```

Comment:
Do we even need to infer things here or can we just use `TRoute extends Route` and then `TRoute['path']['params]`

Reply:
this was required at one point but I can't remember why and changing it doesn't _appear_ to break anything.

Reply:
I don't feel strongly it needs to change. Just seemed odd to not have an extends and to need all the infers.